### PR TITLE
Feature/feature a

### PR DESCRIPTION
**Pull request by ldelbel**
In this project, the goal was to build methods using the Enumerable Module. The following methods were built:

- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count
- my_map (and proc)
- my_inject

### CODE REVIEW

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great start @team :clap: :clap: However, some changes need to be applied in order to get approval for this project. Check :arrow_down: 

- **[REQUIRED]** No debugging instructions in committed code (as discussed [here](https://questions.microverse.org/t/debugging-instructions-code/346)) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/section/debugging_requirement.png) from Pathwright)

- **[REQUIRED]** Comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

### :arrow_right:  my_each, my_each_with_index, my_select

* [x] consider the case when self is a range, a hash

### :arrow_right:  my_map, my_inject

* [x] consider the case when self is a range

##### :arrow_right:  gen

The issues with **my_none?**, **my_any?** and **my_all?** are similar. the methods should be able to: 

* [x] Handle the cases where no block is given (Hint: block_given?)
* [x] Return a suitable boolean result if no arguments were passed.
  ```ruby
	array = [false, true, false, nil, []]
	puts array.my_any? # should print true
	puts array.my_all? # should print false
	puts array.my_none? # should print false
  ```
* [x] Accept primitives, classes (i.e Integer, String, etc...) and Regex Expressions as arguments.
  ```ruby
	words = %w[dog door rod blade]
	puts words.my_all?(/c/) == words.all?(/c/) # false
	puts words.my_any?(/c/) == words.all?(/c/) # false
	puts words.my_none?(/c/) == words.all?(/c/) # true
  ```



### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    elsif test.is_a? Class
+      my_each { |x| check = true if x.class == test }
+    elsif test.is_a? Regexp
+      my_each { |x| check = true if test.match(x) }
+    else
+      test_array = [true, [], {}]
+      my_each { |x| check = true if test_array.include?(x) }
+    end
+    check
+  end
```
Comment (maelfosso) [inline enumerable.rb]:
```rb
2.6.5 :018 > words = %w[dog door rod blade]
 => ["dog", "door", "rod", "blade"] 
2.6.5 :019 > words.any?('cat')
 => false 
2.6.5 :020 > words.my_any?('cat')
 => false 
2.6.5 :021 > ws = words.dup
 => ["dog", "door", "rod", "blade"] 
2.6.5 :022 > ws[0] = 'cat'
 => "cat" 
2.6.5 :023 > ws.any?('cat')
 => true 
2.6.5 :024 > ws.my_any?('cat')
 => false 
2.6.5 :025 > 
```

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      my_each { |x| check = false if yield(x) }
+    elsif test.is_a? Class
+      my_each { |x| check = false if x.class == test }
+    elsif test.is_a? Regexp
+      my_each { |x| check = false if test.match(x) }
+    else
+      my_each { |x| check = false if x == true }
+    end
+    check
+  end
```
Comment (maelfosso) [inline enumerable.rb]:
```rb
2.6.5 :027 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :028 > words.none?(5)
 => true 
2.6.5 :029 > words.my_none?(5)
 => true 
2.6.5 :030 > ws = words.dup
 => ["dog", "door", "rod", "blade"] 
2.6.5 :031 > ws[0] = 5
 => 5 
2.6.5 :032 > ws.none?(5)
 => false 
2.6.5 :033 > ws.my_none?(5)
 => true 
2.6.5 :034 > 
```

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      check
+    elsif test.is_a? Class
+      my_each { |x| check = false if x.class != test }
+    elsif test.is_a? Regexp
+      my_each { |x| check = false unless test.match(x) }
+    else
+      my_each { |x| check = false if x.nil? || x == false }
+    end
+    check
+  end
```
Comment (maelfosso) [inline enumerable.rb]:
```rb
2.6.5 :013 > array
 => [4, 3, 3, 1, 1, 6, 5, 6, 6, 0, 7, 3, 1, 3, 5, 6, 3, 5, 2, 8, 6, 5, 6, 3, 6, 8, 1, 1, 7, 1, 0, 6, 2, 7, 3, 7, 5, 7, 6, 7, 6, 1, 8, 5, 7, 7, 0, 8, 6, 4, 2, 0, 0, 0, 8, 8, 0, 4, 0, 4, 2, 6, 2, 7, 2, 6, 0, 3, 8, 8, 5, 7, 7, 6, 5, 5, 5, 4, 5, 2, 4, 4, 2, 6, 4, 5, 3, 8, 1, 5, 3, 3, 2, 1, 5, 3, 1, 5, 3, 8] 
2.6.5 :014 > array.all?(3)
 => false 
2.6.5 :015 > array.my_all?(3)
 => true 
2.6.5 :016 > 
```

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+  def my_count
+    return size unless block_given?
+
+    j = 0
+    my_each { |x| j += 1 if yield(x) }
+    j
+  end
```
Comment (maelfosso) [inline enumerable.rb]:
```rb
2.6.5 :037 > array
 => [4, 3, 3, 1, 1, 6, 5, 6, 6, 0, 7, 3, 1, 3, 5, 6, 3, 5, 2, 8, 6, 5, 6, 3, 6, 8, 1, 1, 7, 1, 0, 6, 2, 7, 3, 7, 5, 7, 6, 7, 6, 1, 8, 5, 7, 7, 0, 8, 6, 4, 2, 0, 0, 0, 8, 8, 0, 4, 0, 4, 2, 6, 2, 7, 2, 6, 0, 3, 8, 8, 5, 7, 7, 6, 5, 5, 5, 4, 5, 2, 4, 4, 2, 6, 4, 5, 3, 8, 1, 5, 3, 3, 2, 1, 5, 3, 1, 5, 3, 8] 
2.6.5 :038 > array.count(3)
 => 13 
2.6.5 :039 > array.my_count(3)
Traceback (most recent call last):
        5: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        3: from /usr/share/rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):39
        1: from enumerable.rb:79:in `my_count'
ArgumentError (wrong number of arguments (given 1, expected 0))
2.6.5 :040 > 
```

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    results = []
+    if block_given?
+      arr.my_each { |x| results << yield(x) }
+    elsif proc
+      arr.my_each { |x| results << proc.call(x) }
+    else
+      return to_enum
+    end
+    results
+  end
```
Comment (maelfosso) [inline enumerable.rb]:
```rb
2.6.5 :052 > my_proc = proc {|num| num > 10 }
 => #<Proc:0x00000000029401b8@(irb):52> 
2.6.5 :053 > array.my_map(my_proc) {|num| num < 10 } == array.map(&my_proc)
 => false 
2.6.5 :054 > array
 => [4, 3, 3, 1, 1, 6, 5, 6, 6, 0, 7, 3, 1, 3, 5, 6, 3, 5, 2, 8, 6, 5, 6, 3, 6, 8, 1, 1, 7, 1, 0, 6, 2, 7, 3, 7, 5, 7, 6, 7, 6, 1, 8, 5, 7, 7, 0, 8, 6, 4, 2, 0, 0, 0, 8, 8, 0, 4, 0, 4, 2, 6, 2, 7, 2, 6, 0, 3, 8, 8, 5, 7, 7, 6, 5, 5, 5, 4, 5, 2, 4, 4, 2, 6, 4, 5, 3, 8, 1, 5, 3, 3, 2, 1, 5, 3, 1, 5, 3, 8] 
2.6.5 :055 > 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,
Thank you for applying the suggested changes.
You did a great job !!

However, there are still some errors to fix.
Kindly read about them below, then fix them and submit a new code review request

Thanks
Mael FOSSO

**Comment by ldelbel** [comment]
Fixed the problems pointed for my_all?, my_any? and my_none?

But for my_map, kindly check again the test that was made. This is what was given:

2.6.5 :052 > my_proc = proc {|num| num > 10 }
 => #<Proc:0x00000000029401b8@(irb):52> 
2.6.5 :053 > array.my_map(my_proc) {|num| num < 10 } == array.map(&my_proc)
 => false 

Note that in my_proc the test is 'num > 10' (greater than) and in the block given for array.my_map was 'num < 10' (less than). When I did the same test with only .map it returned 'false' as well. Please check again.

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    return to_enum unless block_given?
+
+    arr = [Hash, Range].member?(self.class) ? to_a.flatten : self
+    i = 0
+    while i < arr.length
+      yield(arr[i])
+      i += 1
+    end
+    arr
+  end
```
Comment (MarioDena) [inline enumerable.rb]: - [x] My each should, when a block is given and self is a range returns the range itself after calling the given block once for each element in self. You are returning an array in this scenario we need to return a range.

- [x] Same bad return for when using hashes. When a block is given when self is a hash behaves like #each calls the given block once for each element in self. You need to return a hash you are giving a string back.

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    return to_enum unless block_given?
+
+    arr = [Hash, Range].member?(self.class) ? to_a.flatten : self
+    i = 0
+    while i < arr.length
+      yield(arr[i], i) if arr[i].is_a? Integer
+      i += 1
+    end
+    arr
+  end
```
Comment (MarioDena) [inline enumerable.rb]: Same problems from my_each apply here.

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    elsif test.is_a? Class
+      my_each { |x| check = false if x.class != test }
+    elsif test.is_a? Regexp
+      my_each { |x| check = false unless test.match(x) }
+    else
+      t_ar = [nil, false]
+      my_each { |x| check = false if t_ar.include?(x) || x != test }
+    end
+    check
+  end
```
Comment (MarioDena) [inline enumerable.rb]: - [x] when no block or argument is given returns true when none of the collection members are false or nil. I got a true when a false was expected.

- [x] When a class is passed as an argument returns true if all of the collection is a member of such class::Numeric I got a false here expecting a true.

File: enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    elsif test.is_a? Class
+      my_each { |x| check = true if x.class == test }
+    elsif test.is_a? Regexp
+      my_each { |x| check = true if test.match(x) }
+    else
+      test_array = [true, [], {}]
+      my_each { |x| check = true if test_array.include?(x) || include?(test) }
+    end
+    check
+  end
```
Comment (MarioDena) [inline enumerable.rb]: Same problems from my_all? Maybe they are working the other-way around? Im getting false when expecting true.

File: enumerable.rb [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+      my_each { |x| check = false if yield(x) }
+    elsif test.is_a? Class
+      my_each { |x| check = false if x.class == test }
+    elsif test.is_a? Regexp
+      my_each { |x| check = false if test.match(x) }
+    else
+      my_each { |x| check = false if x == true || include?(test) }
+    end
+    check
+  end
```
Comment (MarioDena) [inline enumerable.rb]: - [x] Same problem here, when no block or argument is given returns true only if none of the collection members is true.

File: enumerable.rb [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    results = []
+    if block_given?
+      arr.my_each { |x| results << yield(x) }
+    elsif proc
+      arr.my_each { |x| results << proc.call(x) }
+    else
+      return to_enum
+    end
+    results
+  end
```
Comment (MarioDena) [inline enumerable.rb]: - [x] The error reported by @maelfosso persist my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

File: enumerable.rb [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+              1
+            end
+      arr.my_each { |x| res = res.send(symbol, x) }
+      return res
+    else
+      arr.my_each { |x| result = result.send(symbol, x) }
+    end
+    result
+  end
+end
```
Comment (MarioDena) [inline enumerable.rb]: - [x] my_inject should raise a "LocalJumpError" when no block or argument is given.

That's all please make the requested adjustments so all your classes behave exactly as the built in counterparts. 

Cheers!

**Review by MarioDena — CHANGES_REQUESTED** [review]
Hello Great work implementing your classes however there are still some changes to do.

## Changes required 

Please check my comments.

**Comment by davidmrtz-dev** [comment]
## Hi. 👨‍🚀
### Very good job so far.

However, since the previous revision, there has been no change, there are still the same issues listed by the previous TSE.   ⚠️ 

****
**Please verify each of them and make the changes required and submit for re-review once you are ready!** 
****
### Status: Changes required. ♻️  
****


**[TSE:  David Elí]** 
Slack: @ David Eli Martinez Garcia

**Review by mikenath223 — CHANGES_REQUESTED** [review]
### Hi Team. 👋

**Nice job here, just a few more steps and we are good to go. 👌**

- [x] There are still issues with my_each and my_each_with_index:
- Both methods, when a range is passed, are expected to return a range but yours returns an array.
- Both methods, when a hash is passed to them, are expected to return a hash but yours returns a string containing the hash keys and values, kindly return a hash back under this condition.

- [x] my_all, when a class is passed, is expected to return true if the array passed to it contains elements that all belong to the class. Yours fails in this example `[1,2,3,4].my_all?(Numeric)`.  Kindly fix it.

- [x] my_any, when a class is passed, is expected to return true if at least one of the elements in the array passed to it belongs to the class. Yours fails in this example `[1,2,3,4].my_any?(Numeric)`.  Kindly fix this also.

- [ ] - my_map is expected to run only the proc when both a block and a proc are passed to it but it seems yours isn't structured to take a single argument as it throws an error of `wrong number of arguments given 1 expected 0`.

- [x] In addition, the live link that exists in your readMe currently leads nowhere, kindly remove it. Sorry, this wasn't mentioned in the previous reviews.

Kindly fix these issues and request another review👍
Feel free to contact me on slack(michgolden) if you need further clarification.

🚀🚀HAPPY CODING🚀🚀

**Comment by lekegitrepo** [comment]
### Good job :+1: 
#### This project is approved :100: :1st_place_medal:

#### _Reviewed by Temitayo Adeleke._

**Review by lekegitrepo — APPROVED** [review]
#### Status: Approved.

**Pull request merged**